### PR TITLE
created program list and solution list API's

### PR DESCRIPTION
**Pull request by deepakr123**

### CODE REVIEW

File: common/cassandra_func.js [hunk 1]
```diff
@@ -2,16 +2,7 @@ var db = require('../db');
 
 let checkReqInCassandra = async function (reqBody, callback) {
     reqBody1 = JSON.stringify(reqBody)
```
Comment (aks30) [inline common/cassandra_func.js]: @deepakr123 - No need for another variable. Please remove

File: migrations/20200608133617-update_list_observation_solutions_query.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+module.exports = {
+  async up(db) {
+    global.migrationMsg = "Update list observation solutions query";
+    
+    
+    if (!cassandra) {
+      throw new Error("Cassandra connection not available.");
+    }
+
+    const query = 'SELECT id FROM reports.druidqueries WHERE qid = ? ALLOW FILTERING';
```
Comment (aks30) [inline migrations/20200608133617-update_list_observation_solutions_query.js]: @deepakr123 - Take table name from env or config.json, for this and all other migrations

File: migrations/20200608133617-update_list_observation_solutions_query.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    const result = await cassandra.execute(query, [ 'list_observation_solutions_query' ], { prepare: true });
```
Comment (aks30) [inline migrations/20200608133617-update_list_observation_solutions_query.js:14]: @deepakr123 - Let's have 2 main queries, programs_list_query and solution_list_query which is common for observation, institutional and individual assessment

File: migrations/20200610113619-update_list_assessment_programs_query.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    const query = 'SELECT id FROM reports.druidqueries WHERE qid = ? ALLOW FILTERING';
+    const result = await cassandra.execute(query, [ 'list_assessment_programs_query' ], { prepare: true });
+    const row = result.rows;
+   
+   
+    let queries = 
+      [{
+        query: 'UPDATE reports.druidqueries SET query=? WHERE id=?',
+        params: ['{"queryType":"groupBy","dataSource":"sl_assessment_dev","granularity":"all","dimensions":["programName","programId","programDescription","programExternalId","solutionName","solutionId","solutionDescription","solutionExternalId"],"filter":{"type":"and","fields":[{"type":"selector","dimension":"","value":""},{"type":"or","fields":[{"type":"and","fields":[{"type":"selector","dimension":"userId","value":""},{"type":"selector","dimension":"isAPrivateProgram","value":true}]},{"type":"selector","dimension":"isAPrivateProgram","value":false}]}]},"aggregations":[],"postAggregations":[],"intervals":["1901-01-01T00:00:00+00:00/2101-01-01T00:00:00+00:00"]}',row[0].id]
```
Comment (aks30) [inline migrations/20200610113619-update_list_assessment_programs_query.js]: @deepakr123 - let's use the name as sl_assessment

File: controllers/v1/assessments.js [hunk 5]
```diff
@@ -49,15 +49,23 @@ exports.listPrograms = async function (req, res) {
...
+
+                let bodyParam = JSON.parse(result.query);
+
                 if (config.druid.assessment_datasource_name) {
                     bodyParam.dataSource = config.druid.assessment_datasource_name;
                 }
-                bodyParam.filter.dimension = req.body.entityType;
-                bodyParam.filter.value = req.body.entityId;
+               
+                let userId = await getCreatedByField(req, res);
```
Comment (aks30) [inline controllers/v1/assessments.js]: @deepakr123 - Please evaluate if we can make createdBy as an array.

File: controllers/v1/solutions.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            result: false,
+            message: 'entityId,entityType,programId and type are required fields'
+        }
+        res.send(response);
+    }
+    else {
+
+        let query;
+
+        if (req.body.reportType == "my") {
```
Comment (aks30) [inline controllers/v1/solutions.js]: @deepakr123 - Please evaluate of both responses can be combined into 1 response.

File: common/cassandra_func.js (hunk 1 above)
Comment (deepakr123) [reply common/cassandra_func.js]: resolved

File: migrations/20200608133617-update_list_observation_solutions_query.js (hunk 2 above)
Comment (deepakr123) [reply migrations/20200608133617-update_list_observation_solutions_query.js]: resolved

File: controllers/v1/programs.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+                    message: 'INTERNAL_SERVER_ERROR'
+                }
+                res.send(response);
+            })
+        }
+};
+
+
+// Function for getting createdBy field from header access token
+async function getCreatedByField(req, res) {
```
Comment (aks30) [inline controllers/v1/programs.js]: @deepakr123 - No need to validate with keycloak, set it in the request body after request is autheticated in the middleware.

File: controllers/v1/programs.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+             }
+             
+             bodyParam.filter.fields[1].fields[0].fields[0].dimension = "userId";
+             bodyParam.filter.fields[1].fields[0].fields[0].value = id;
+
+             let options = config.druid.options;
+             options.method = "POST";
+             options.body = bodyParam;
+
+             let assessmentPrograms = await rp(options);
```
Comment (aks30) [inline controllers/v1/programs.js]: @deepakr123 - Run queries in both data sources in parallel in all the files

File: controllers/v1/solutions.js [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+            message: 'entityId,entityType,programId are required fields'
+        }
+        res.send(response);
+    }
+    else {
+
+          //get the createdBy field
+          let id = await getCreatedByField(req, res);
+
+          let myAssessmentSolutions = await mySolutionsList(req, id, filesHelper.assessment);
```
Comment (aks30) [inline controllers/v1/solutions.js]: @deepakr123 - All 4 queries can be in parallel.

File: controllers/v1/assessments.js (hunk 5 above)
Comment (deepakr123) [reply controllers/v1/assessments.js]: resolved

File: migrations/20200608133617-update_list_observation_solutions_query.js (hunk 3 above)
Comment (deepakr123) [reply migrations/20200608133617-update_list_observation_solutions_query.js:14]: resolved

File: migrations/20200610113619-update_list_assessment_programs_query.js (hunk 4 above)
Comment (deepakr123) [reply migrations/20200610113619-update_list_assessment_programs_query.js]: resolved

File: controllers/v1/solutions.js (hunk 6 above)
Comment (deepakr123) [reply controllers/v1/solutions.js]: resolved

File: controllers/v1/programs.js (hunk 7 above)
Comment (deepakr123) [reply controllers/v1/programs.js]: resolved

File: controllers/v1/programs.js (hunk 8 above)
Comment (deepakr123) [reply controllers/v1/programs.js]: resolved

File: controllers/v1/solutions.js (hunk 9 above)
Comment (deepakr123) [reply controllers/v1/solutions.js]: resolved

**Pull request merged**